defs/go.bzl

```diff
@@ -12,87 +12,87 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-load("@io_bazel_rules_go//go:def.bzl", "GoLibrary", "GoPath", "go_context", "go_path", "go_rule")
+load("@bazel_skylib//lib:paths.bzl", "paths")
+load("@io_bazel_rules_go//go:def.bzl", "GoArchive", "GoLibrary", "go_context", "go_path", "go_rule")
 
-def _compute_genrule_variables(resolved_srcs, resolved_outs, dep_import_paths):
+def _compute_genrule_variables(srcs, outs):
+    resolved_srcs = [src.path for src in srcs]
+    resolved_outs = [out.path for out in outs]
     variables = {
-        "SRCS": " ".join([src.path for src in resolved_srcs]),
-        "OUTS": " ".join([out.path for out in resolved_outs]),
-        "GO_IMPORT_PATHS": " ".join(dep_import_paths),
+        "SRCS": " ".join(resolved_srcs),
+        "OUTS": " ".join(resolved_outs),
     }
     if len(resolved_srcs) == 1:
-        variables["<"] = resolved_srcs[0].path
-    if len(resolved_outs) == 1:
-        variables["@"] = resolved_outs[0].path
+        variables["<"] = resolved_srcs[0]
+    if len(outs) == 1:
+        variables["@"] = resolved_outs[0]
     return variables
 
 def _go_genrule_impl(ctx):
     go = go_context(ctx)
 
-    transitive_depsets = []
-    label_dict = {}
-    go_paths = []
+    transitive_libs = depset()
+    for dep in ctx.attr.go_deps:
+        transitive_libs = depset(transitive = [transitive_libs, dep[GoArchive].transitive])
 
-    for dep in ctx.attr.srcs:
-        transitive_depsets.append(dep.files)
-        label_dict[dep.label] = dep.files.to_list()
+    gopath = []
+    for lib in transitive_libs.to_list():
+        for srcfile in lib.srcs:
+            target = ctx.actions.declare_file(paths.join(
+                "gopath/src",
+                lib.importpath,
+                paths.basename(srcfile.path),
+            ))
 
-    dep_import_paths = []
-    for dep in ctx.attr.go_deps:
-        dep_import_paths.append(dep[GoLibrary].importpath)
+            args = ctx.actions.args()
+            args.add(srcfile.path)
+            args.add(target.path)
 
-    for go_path in ctx.attr.go_paths:
-        transitive_depsets.append(go_path.files)
-        label_dict[go_path.label] = go_path.files.to_list()
+            ctx.actions.run(
```

Comment:
This generates a lot of actions. It's not slow (e.g. takes a couple seconds to setup the gopath for the main openapi generation, which has a massive number of dependencies), but it cold be made faster.